examples/rust/components/blobby/src/lib.rs

```
@@ -121,251 +77,165 @@ impl Guest for Blobby {
+    let body = IncomingValue::incoming_value_consume_async(incoming)?;
+    debug!("successfully got object stream");
+    Ok((body, metadata.size))
 }
 
+/// Write the object data to the object in the container
 fn put_object(
-    container_name: &String,
+    container: Container,
     object_name: &String,
-    data: InputStream,
-    mut content_length: u64,
-) -> Result<StatusCode> {
-    let container =
-        blobstore::blobstore::get_container(container_name).map_err(Error::from_blobstore_error)?;
-    let result_value = blobstore::types::OutgoingValue::new_outgoing_value();
-
-    let stream = result_value
+    mut data: http::IncomingBody,
+    content_length: u64,
+) -> Result<(), String> {
+    let result_value = OutgoingValue::new_outgoing_value();
+    let mut stream = result_value
         .outgoing_value_write_body()
         .expect("failed to get outgoing value output stream");
-
     if let Err(e) = container.write_data(object_name, &result_value) {
-        log(
-            Level::Error,
-            "put_object",
-            &format!("Failed to write data to blobstore: {}", e),
-        );
-        return Err(Error {
-            status_code: StatusCode::INTERNAL_SERVER_ERROR,
-            message: format!("Failed to write data to blobstore: {}", e),
-        });
+        error!("Failed to write data to blobstore: {e}");
+        return Err(format!("Failed to write data to blobstore: {e}"));
     }
-    while content_length > 0 {
-        let len = stream
-            .blocking_splice(&data, content_length)
-            .expect("failed to stream data from http response to blobstore");
-        content_length = content_length.saturating_sub(len);
+
+    let copied_bytes = std::io::copy(&mut data, &mut stream).map_err(|e| e.to_string())?;
+    if copied_bytes != content_length {
+        warn!("Expected to copy {content_length} bytes, but only copied {copied_bytes} bytes");
```

Nit: Should this return an error instead of just logging a warning?

---

I'm not really sure, I think that copying fewer bytes could happen if the process disconnects but copying more bytes than expected would be bad, it's just with `io::copy` we don't limit the bytes anymore 🤔

---

It is probably good to return an error either way then so the caller can decide what to do